packages/vite/src/node/plugins/assetImportMetaUrl.ts

```diff
@@ -120,11 +93,39 @@ export function assetImportMetaUrlPlugin(config: ResolvedConfig): Plugin {
             )
             builtUrl = url
           }
-          s.update(
-            index,
-            index + exp.length,
-            `new URL(${JSON.stringify(builtUrl)}, self.location)`,
-          )
+
+          // potential dynamic template string
+          if (rawUrl[0] === '`' && /\$\{/.test(rawUrl)) {
+            const ast = this.parse(rawUrl)
+            const templateLiteral = (ast as any).body[0].expression
+            if (templateLiteral.expressions.length) {
+              const pattern = buildGlobPattern(templateLiteral)
+              if (pattern.startsWith('**')) {
+                // don't transform for patterns like this
+                // because users won't intend to do that in most cases
+                continue
+              }
+
+              // Note: native import.meta.url is not supported in the baseline
+              // target so we use the global location here. It can be
+              // window.location or self.location in case it is used in a Web Worker.
+              // @see https://developer.mozilla.org/en-US/docs/Web/API/Window/self
+              s.update(
+                index,
+                index + exp.length,
+                `new URL(
+                   import.meta.glob(${JSON.stringify(pattern)},
+                   { eager: true, import: 'default', as: 'url' }
+                 )[${rawUrl}] || \`${builtUrl}\`, self.location)`,
+              )
```

Comment:
This is a clever way to support the old behavior. But needing to do this for all cases seems wrong to me. The generated code would not be clean. I think that https://github.com/vitejs/vite/issues/11157 may not be an issue, but we need to clarify if needed the docs to add a vite-ignore in this case. @antfu what do you think?

Reply:
To me personally, I feel the statement of `return new URL(`./dir/${name}.png`, import.meta.url).href` making Vite bundle all png files under that directory is a bit too implicit. So I also consider this is not a good solution. To me, I feel we should resolve it to the correct path like v3.1 but not maybe not bundle all the assets by default.

Reply:
In version 3.1, the `assetImportMetaUrlPlugin` was not introduced in dev mode, so `undefined` will not appear.

Reply:
but in build mode, `undefined` will appear.

Reply:
> This is a clever way to support the old behavior. But needing to do this for all cases seems wrong to me. The generated code would not be clean. I think that #11157 may not be an issue, but we need to clarify if needed the docs to add a vite-ignore in this case. @antfu what do you think?

Do you mean adding the logic to handle `vite-ignore` in `assetImportMetaUrlPlugin`?